### PR TITLE
LibJS: Add spec comments to ArrayBuffer.prototype

### PR DESCRIPTION
**Pull request by jamierocks**

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 1]
```diff
@@ -37,51 +38,96 @@ ArrayBufferPrototype::~ArrayBufferPrototype()
 // 25.1.5.3 ArrayBuffer.prototype.slice ( start, end ), https://tc39.es/ecma262/#sec-arraybuffer.prototype.slice
 JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::slice)
 {
+    // 1. Let O be the this value.
+    // 2. Perform ? RequireInternalSlot(O, [[ArrayBufferData]]).
     auto* array_buffer_object = TRY(typed_this_value(global_object));
-    // FIXME: Check for shared buffer
+
+    // 3. If IsSharedArrayBuffer(O) is true, throw a TypeError exception.
+    // fixme: shared buffer
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:46]: Please capitalise FIXME, and be a bit more verbose - comments should be sentences.

Either:

```
// 3. If IsSharedArrayBuffer(O) is true, throw a TypeError
// FIXME: Handle SharedArrayBuffer.
```

or

```
// FIXME:
// 3. If IsSharedArrayBuffer(O) is true, throw a TypeError
```

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 2]
```diff
@@ -37,51 +38,96 @@ ArrayBufferPrototype::~ArrayBufferPrototype()
...
+    // 5. Let len be O.[[ArrayBufferByteLength]].
     auto length = array_buffer_object->byte_length();
 
+    // 6. Let relativeStart be ? ToIntegerOrInfinity(start).
     auto relative_start = TRY(vm.argument(0).to_integer_or_infinity(global_object));
 
     double first;
-    if (relative_start < 0)
+    if (Value(relative_start).is_negative_infinity()) {
+        // 7. If relativeStart is -∞, let first be 0.
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:60]: This goes above the if

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 3]
```diff
@@ -37,51 +38,96 @@ ArrayBufferPrototype::~ArrayBufferPrototype()
...
+        first = 0;
+    } else if (relative_start < 0) {
+        // 8. Else if relativeStart < 0, let first be max(len + relativeStart, 0).
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:63]: Likewise

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 4]
```diff
@@ -37,51 +38,96 @@ ArrayBufferPrototype::~ArrayBufferPrototype()
...
         first = max(length + relative_start, 0.0);
-    else
+    } else {
+        // 9. Else, let first be min(relativeStart, len).
         first = min(relative_start, (double)length);
+    }
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:68]: No curly braces needed here with the comments moved.

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 5]
```diff
@@ -37,51 +38,96 @@ ArrayBufferPrototype::~ArrayBufferPrototype()
...
 
+    // 10. If end is undefined, let relativeEnd be len; else let relativeEnd be ? ToIntegerOrInfinity(end).
     auto relative_end = vm.argument(1).is_undefined() ? length : TRY(vm.argument(1).to_integer_or_infinity(global_object));
 
     double final;
-    if (relative_end < 0)
+    if (Value(relative_end).is_negative_infinity()) {
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:74]: Same here

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 6]
```diff
@@ -37,51 +38,96 @@ ArrayBufferPrototype::~ArrayBufferPrototype()
...
     auto new_length = max(final - first, 0.0);
 
+    // 15. Let ctor be ? SpeciesConstructor(O, %ArrayBuffer%).
     auto* constructor = TRY(species_constructor(global_object, *array_buffer_object, *global_object.array_buffer_constructor()));
+
+    // 16. Let new be ? Construct(ctor, « 𝔽(newLen) »).
     auto* new_array_buffer = TRY(construct(global_object, *constructor, Value(new_length)));
 
+    // 17. Perform ? RequireInternalSlot(new, [[ArrayBufferData]]).
+    // fixme: is this the exception we want here?
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:95]: yes?

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 7]
```diff
@@ -37,51 +38,96 @@ ArrayBufferPrototype::~ArrayBufferPrototype()
...
     if (!is<ArrayBuffer>(new_array_buffer))
         return vm.throw_completion<TypeError>(global_object, ErrorType::SpeciesConstructorDidNotCreate, "an ArrayBuffer");
     auto* new_array_buffer_object = static_cast<ArrayBuffer*>(new_array_buffer);
 
-    // FIXME: Check for shared buffer
+    // 18. If IsSharedArrayBuffer(new) is true, throw a TypeError exception.
+    // fixme: check for shared buffer
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:101]: Same here

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 8]
```diff
@@ -37,51 +38,96 @@ ArrayBufferPrototype::~ArrayBufferPrototype()
...
 
+    // 22. NOTE: Side-effects of the above steps may have detached O.
+    // 23. If IsDetachedBuffer(O) is true, throw a TypeError exception.
     if (array_buffer_object->is_detached())
         return vm.throw_completion<TypeError>(global_object, ErrorType::DetachedArrayBuffer);
 
-    // This is ugly, is there a better way to do this?
-    array_buffer_object->buffer().span().slice(first, new_length).copy_to(new_array_buffer_object->buffer().span());
+    // 24. Let fromBuf be O.[[ArrayBufferData]].
+    auto from_buf = array_buffer_object->buffer();
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:121]: Let's not create a copy?!

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 9]
```diff
@@ -37,51 +38,96 @@ ArrayBufferPrototype::~ArrayBufferPrototype()
...
+
+    // 25. Let toBuf be new.[[ArrayBufferData]].
+    auto to_buf = new_array_buffer_object->buffer();
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:124]: Same here, don't think this actually works. If we copy the buffer, the returned object is unaffected.

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 10]
```diff
@@ -134,12 +134,22 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::slice)
 // 25.1.5.1 get ArrayBuffer.prototype.byteLength, https://tc39.es/ecma262/#sec-get-arraybuffer.prototype.bytelength
 JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::byte_length_getter)
 {
+    // 1. Let O be the this value.
+    // 2. Perform ? RequireInternalSlot(O, [[ArrayBufferData]]).
     auto* array_buffer_object = TRY(typed_this_value(global_object));
-    // FIXME: Check for shared buffer
+
+    // 3. If IsSharedArrayBuffer(O) is true, throw a TypeError exception.
+    // fixme: check for shared buffer
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:142]: Same here

**Review by linusg — CHANGES_REQUESTED** [review]
This breaks tests.

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp (hunk 9 above)
Comment (jamierocks) [reply Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:124]: Silly me, this is one of those last minute stylistic changes I made - but never ran test-js for.

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp (hunk 1 above)
Comment (jamierocks) [reply Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:46]: I had meant to replace this for `// fixme: check for shared buffer` as elsewhere, I'll replace with the appropriate capitalisation also :)

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp (hunk 6 above)
Comment (jamierocks) [reply Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:95]: I had meant to provide a comment here, or on Discord on this, I got drawn away by something else after submitting the pr.

My point here is, in the spec both step 2 and 17 are performing `RequireInternalSlot(new, [[ArrayBufferData]])` - however are throwing two different exceptions here.
- Step 1 throws `vm.throw_completion<TypeError>(global_object, ErrorType::NotAnObjectOfType, PrototypeType::display_name())`
- Step 17 throws `vm.throw_completion<TypeError>(global_object, ErrorType::SpeciesConstructorDidNotCreate, "an ArrayBuffer")`

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp (hunk 2 above)
Comment (jamierocks) [reply Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:60]: I'll amend that - I wasn't a hundred on this, but followed what TypedArrayPrototype was doing :)

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 11]
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2020, Linus Groh <linusg@serenityos.org>
+ * Copyright (c) 2021-2022, Jamie Mansfield <jmansfield@cadixdev.org>
```
Comment (IdanHo) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:3]: Did you work on this file in 2021? :^)

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp (hunk 6 above)
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp]: They check slightly different things though :)

The first is for the this value, the second for the object created by the constructor returned from `SpeciesConstructor`, so I think it's ok to throw different errors.

**Review by linusg — APPROVED** [review]
Thanks! :^)

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp (hunk 11 above)
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:3]: Well now I have egg on my face! LGTM then :^)

**Pull request merged**